ddtrace/contrib/rq/__init__.py

```diff
@@ -0,0 +1,236 @@
+    worker_service_name=get_env('rq', 'worker_service_name', default='rq-worker'),
+    app='rq',
+    app_type=AppTypes.worker,
+    distributed_tracing_enabled=False,
+))
+
+
+def with_instance_pin(func):
+    """Helper to wrap a function wrapper and ensure an enabled pin is available for the `instance`"""
+    def with_cls(cls):
+        def wrapper(wrapped, instance, args, kwargs):
+            # DEV: this will be replaced with the module given when patching on import
+            import rq
+
+            # DEV: precedence when finding Pin is instance, class, rq module in that order
+            pin = Pin._find(instance, cls, rq)
+            if not pin or not pin.enabled():
+                return wrapped(*args, **kwargs)
+            return func(rq, pin, wrapped, instance, args, kwargs)
+        return wrapper
+    return with_cls
+
+
+propagator = HTTPPropagator()
+
+
+@with_instance_pin
+def traced_queue_enqueue_job(rq, pin, func, instance, args, kwargs):
+    job = args[0]
+
+    with pin.tracer.trace('rq.queue.enqueue_job', service=pin.service, resource=job.func_name) as span:
+        span.set_tag('queue.name', instance.name)
+        span.set_tag('job.id', job.get_id())
+        span.set_tag('job.func_name', job.func_name)
+
+        # If the queue is_async then add distributed tracing headers to the job
+        if instance.is_async and config.rq['distributed_tracing_enabled']:
+            propagator.inject(span.context, job.meta)
+
+        try:
+            return func(*args, **kwargs)
+        finally:
+            # If the queue is not async then the job is run immediately so we can
+            # report the results
+            if not instance.is_async:
+                span.set_tag('job.status', job.get_status())
+
+
+@with_instance_pin
+def traced_queue_fetch_job(rq, pin, func, instance, args, kwargs):
+    with pin.tracer.trace('rq.queue.fetch_job', service=pin.service) as span:
+        span.set_tag('job.id', args[0])
+        return func(*args, **kwargs)
+
+
+@with_instance_pin
+def traced_perform_job(rq, pin, func, instance, args, kwargs):
+    """Trace rq.Worker.perform_job
+    """
+    # `perform_job` is executed in a freshly forked, short-lived instance
+    job = args[0]
+
+    ctx = propagator.extract(job.meta)
+    if ctx.trace_id:
+        pin.tracer.context_provider.activate(ctx)
+
+    try:
+        with pin.tracer.trace('rq.worker.perform_job', service=pin.service, resource=job.func_name) as span:
+            span.set_tag('job.id', job.get_id())
+            return func(*args, **kwargs)
+    finally:
+        if job.get_status() == rq.job.JobStatus.FAILED:
+            span.error = 1
+            span.set_tag('exc_info', job.exc_info)
+        span.set_tag('status', job.get_status())
+        span.set_tag('origin', job.origin)
+
+        # DEV: force flush to agent since the process `os.exit()`s
+        #      immediately after this method returns
+        pin.tracer.writer.stop()
```

Comment:
I think it's ok for now but this is a big hint that the `Tracer` API is not good enough.